src/backend/mpi/swdsm.cpp

```diff
@@ -436,315 +433,262 @@ void handler(int sig, siginfo_t *si, void *unused){
+	return gptr.offset();
+}
 
-	if(loadtag>=size_of_all){//Trying to access/prefetch out of memory
-		return;
-	}
-	homenode = getHomenode(loadtag);
-	unsigned long cacheIndex = loadline;
-	if(cacheIndex >= cachesize){
-		printf("idx > size   cacheIndex:%ld cachesize:%ld\n",cacheIndex,cachesize);
-		return;
-	}
-	sem_wait(&ibsem);
+std::size_t peek_offset(std::size_t addr) {
+	dd::global_ptr<char> gptr(reinterpret_cast<char*>(
+			addr + reinterpret_cast<unsigned long>(startAddr)), false, false);
+	return gptr.peek_offset();
+}
 
+void load_cache_entry(std::size_t aligned_access_offset) {
 
-	unsigned long pageAddr = loadtag;
-	unsigned long blocksize = pagesize*CACHELINE;
-	unsigned long lineAddr = pageAddr/blocksize;
-	lineAddr *= blocksize;
+	/* If it's not an ArgoDSM address, do not handle it */
+	if(aligned_access_offset >= size_of_all){
+		return;
+	}
 
-	unsigned long startidx = cacheIndex/CACHELINE;
-	startidx*=CACHELINE;
-	unsigned long end = startidx+CACHELINE;
+	/* Assign node bit IDs */
+	const std::uintptr_t node_id_bit = 1 << getID();
+	const std::uintptr_t node_id_inv_bit = ~node_id_bit;
 
-	if(end>=cachesize){
-		end = cachesize;
-	}
+	/* Calculate start values and store some parameters */
+	const std::size_t block_size = pagesize*CACHELINE;
+	const std::size_t cache_index = getCacheIndex(aligned_access_offset);
+	const std::size_t start_index = align_backwards(cache_index, CACHELINE);
+	std::size_t end_index = start_index+CACHELINE;
+	const argo::node_id_t load_node = get_homenode(aligned_access_offset);
+	const std::size_t load_offset = get_offset(aligned_access_offset);
 
-	argo_byte tmpstate = cacheControl[startidx].state;
-	unsigned long tmptag = cacheControl[startidx].tag;
+	sem_wait(&ibsem);
 
-	if(tmptag == lineAddr && tmpstate != INVALID){
+	/* Return if requested cache entry is already up to date. */
+	if(cacheControl[start_index].tag == aligned_access_offset &&
+			cacheControl[start_index].state != INVALID){
 		sem_post(&ibsem);
 		return;
 	}
 
+	/* Adjust end_index to ensure the whole chunk to fetch is on the same node */
+	for(std::size_t i = start_index+CACHELINE, p = CACHELINE;
+					i < start_index+load_size;
+					i+=CACHELINE, p+=CACHELINE){
+		const std::size_t temp_addr = aligned_access_offset + p*block_size;
+		/* Increase end_index if it is within bounds and on the same node */
+		if(temp_addr < size_of_all && i < cachesize){
+			const argo::node_id_t temp_node = peek_homenode(temp_addr);
+			const std::size_t temp_offset = peek_offset(temp_addr);
+			if(temp_node == load_node && temp_offset == (load_offset + p*block_size)){
+				end_index+=CACHELINE;
+			}else{
+				break;
+			}
+		}else{
+			/* Stop when either condition is not satisfied */
+			break;
+		}
+	}
 
-	void * lineptr = (char*)startAddr + lineAddr;
-
-	if(cacheControl[startidx].tag  != lineAddr){
-		if(cacheControl[startidx].tag  != lineAddr){
+	/* Allocate space for loading things */
```

Comment:
Either make this comment more precise (there is no space allocation, it computes some sizes; what are the "things", etc.), or take it out.